Remove req_params facility from API

xemul opened the pull request:
The class was introduced to facilitate path and query parameters parsing from requests, but in fact it's mostly dead code.

First, the class introduces the concept of  "mandatory" parameters which are seastar path params. If missing, the parameter validation throws, but in all cases where this option is used in scylla it's impossible to get empty path param -- if the parameter is missing seastar returns 404 (not found) before calling handler.

Second, the req_params::get<T>() doesn't work for anything but string argument (or types such that optional<T> can be implicitly casted to optional<sstring>). And it's in fact only used to get sstrings, so it compiles and works so far.

The remaining ability to parse bool from string is partially duplicated by the validate_bool() method. Using plain method to parse string to bool is less code than req_params introduce.

One (arguably) useful thing req_params do it validate the incoming request _not_ to contain unknown query parameters. However, quite a few endpoints use this, most of them just cherry-pick parameters they want and ignore the others. There's already a comprehensive description of accepted parameters for each endpoint in api-doc/ and req_params duplicate it. Good validation code should rely on api-doc/, not on its partial copy.

Having said that, this PR introduces validate_bool_x() helper to do req_params-like parsing of strings to bools, patches existing handlers to use existing parameters parsing facilities (such as validate_keyspace() and parse_table_infos()) and drops the req_params.

avikivity commented:
@bhalevy you added this class, so please review

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17267/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17267/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5205/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8372/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16455/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17267/)
#### Failed Tests (10/44142):
- [test_topology_ops_encrypted[True].dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17267/testReport/junit/cluster/test_topology_ops_encrypted/Tests___Unit_Tests___test_topology_ops_encrypted_True__dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_ops_encrypted[True].dev.3)
- [test_storage_service_keyspace_bad_param.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17267/testReport/junit/rest_api/test_storage_service/Tests___Unit_Tests___test_storage_service_keyspace_bad_param_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_keyspace_bad_param.debug.1)
- [test_storage_service_keyspace_bad_param.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17267/testReport/junit/rest_api/test_storage_service/Tests___Unit_Tests___test_storage_service_keyspace_bad_param_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_keyspace_bad_param.debug.2)
- [test_storage_service_keyspace_bad_param.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17267/testReport/junit/rest_api/test_storage_service/Tests___Unit_Tests___test_storage_service_keyspace_bad_param_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_keyspace_bad_param.debug.3)
- [test_storage_service_keyspace_bad_param.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17267/testReport/junit/rest_api/test_storage_service/Tests___Unit_Tests___test_storage_service_keyspace_bad_param_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_keyspace_bad_param.dev.1)
- [test_storage_service_keyspace_bad_param.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17267/testReport/junit/rest_api/test_storage_service/Tests___Unit_Tests___test_storage_service_keyspace_bad_param_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_keyspace_bad_param.dev.2)
- [test_storage_service_keyspace_bad_param.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17267/testReport/junit/rest_api/test_storage_service/Tests___Unit_Tests___test_storage_service_keyspace_bad_param_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_keyspace_bad_param.dev.3)
- [test_storage_service_keyspace_bad_param.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17267/testReport/junit/rest_api/test_storage_service/Tests___Unit_Tests___test_storage_service_keyspace_bad_param_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_keyspace_bad_param.release.1)
- [test_storage_service_keyspace_bad_param.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17267/testReport/junit/rest_api/test_storage_service/Tests___Unit_Tests___test_storage_service_keyspace_bad_param_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_keyspace_bad_param.release.2)
- [test_storage_service_keyspace_bad_param.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17267/testReport/junit/rest_api/test_storage_service/Tests___Unit_Tests___test_storage_service_keyspace_bad_param_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_storage_service_keyspace_bad_param.release.3)
#### Build Details:
- Duration: 11 hr
- Builder: spider11.cloudius-systems.com

bhalevy commented on api/storage_service.cc:
```diff
@@ -127,6 +127,21 @@ bool validate_bool(const sstring& param) {
     }
 }
 
+bool validate_bool_x(const sstring& param, bool default_value) {
```
Since it has a different signature, it can be simply called `validate_bool` as well.
And if the default_value would be an optional, default initialized to std::nullopt, and the return value be an optional too, maybe both functions can be unified

xemul replied:
Agree. However, I found 8 different ways of parsing a string into a bool in our API handlers. I planned to generalize it to some extent as a followup.

bhalevy commented on api/column_family.cc:
```diff
@@ -1074,19 +1074,12 @@ void set_column_family(http_context& ctx, routes& r, sharded<db::system_keyspace
-            std::pair("flush_memtables", mandatory::no),
-            std::pair("consider_only_existing_data", mandatory::no),
-            std::pair("split_output", mandatory::no),
-        });
```
I actually like the data-defined structure for the api parameters better than the code-based approach since it allows one to clearly understand the interface in a single glance rather than following custom parsing and processing for each parameter.

What is the motivation for getting rid of req_params?

xemul replied:
> What is the motivation for getting rid of req_params? 

It's in the PR description -- the req_params is mostly dead code, replacing it with plain `req->get_(query|path)_param()` and the newly introduced `validate_bool_x()` gives the same result with much times less code (the whole set -141:+39)

xemul replied:
With one exception -- the req_params::process() captures "unknown" parameters in the request and returns back 400(Bad Request). This part is dropped by this PR. However:

1. there are only 5 endpoints that work like that and 1 (single) test that checks it
2. req_params duplicates description from api/api-doc/

So I suggest to re-introddce incoming request validation as generic code that's based on api-doc data

xemul commented:
upd:
- spellcheck fixes
- removed a test that validated "unrecognized" parameters for storage_service/scrub/ endpoint (and updated PR description with the reasoning)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17279/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17279/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17279/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: rest_api/test_storage_service
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16472/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5221/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8389/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17279/)
#### Build Details:
- Duration: 9 hr 15 min
- Builder: spider9.cloudius-systems.com

xemul commented:
@scylladb/scylla-maint , please consider merging
